lib/ContextResolver.js

```diff
@@ -0,0 +1,243 @@
+      context = context['@context'];
+    }
+
+    // context is one or more contexts
+    if(!_isArray(context)) {
+      context = [context];
+    }
+
+    // resolve each context in the array
+    const allResolved = [];
+    for(const ctx of context) {
+      if(_isString(ctx)) {
+        // see if `ctx` has been resolved before...
+        let resolved = this._get(ctx);
+        if(!resolved) {
+          // not resolved yet, resolve
+          resolved = await this._resolveRemoteContext(
+            {url: ctx, documentLoader, base, cycles});
+        }
+
+        // add to output and continue
+        if(_isArray(resolved)) {
+          allResolved.push(...resolved);
+        } else {
+          allResolved.push(resolved);
+        }
+        continue;
+      }
+      if(ctx === null) {
+        // handle `null` context, nothing to cache
+        allResolved.push(new ResolvedContext({document: null}));
+        continue;
+      }
+      if(!_isObject(ctx)) {
+        _throwInvalidLocalContext(context);
+      }
+      // context is an object, get/create `ResolvedContext` for it
+      const key = JSON.stringify(ctx);
+      let resolved = this._get(key);
+      if(!resolved) {
+        // create a new static `ResolvedContext` and cache it
+        resolved = new ResolvedContext({document: ctx});
+        this._cacheResolvedContext({key, resolved, tag: 'static'});
+      }
+      allResolved.push(resolved);
+    }
+
+    return allResolved;
+  }
+
+  _get(key) {
+    // get key from per operation cache; no `tag` is used with this cache so
+    // any retrieved context will always be the same during a single operation
+    let resolved = this.perOpCache.get(key);
+    if(!resolved) {
+      // see if the shared cache has a `static` entry for this URL
+      const tagMap = this.sharedCache.get(key);
+      if(tagMap) {
+        resolved = tagMap.get('static');
+        if(resolved) {
+          this.perOpCache.set(key, resolved);
+        }
+      }
+    }
+    return resolved;
+  }
+
+  _cacheResolvedContext({key, resolved, tag}) {
+    this.perOpCache.set(key, resolved);
+    if(tag !== undefined) {
+      let tagMap = this.sharedCache.get(key);
+      if(!tagMap) {
+        tagMap = new Map();
+        this.sharedCache.set(key, tagMap);
+      }
+      tagMap.set(tag, resolved);
+    }
+    return resolved;
+  }
+
+  async _resolveRemoteContext({url, documentLoader, base, cycles}) {
+    // resolve relative URL and fetch context
+    url = prependBase(base, url);
+    const {context, remoteDoc} = await this._fetchContext(
+      {url, documentLoader, cycles});
+
+    // update base according to remote document and resolve any relative URLs
+    base = remoteDoc.documentUrl || url;
+    _resolveContextUrls({context, base});
+
+    // resolve, cache, and return context
+    const resolved = await this.resolve(
+      {context, documentLoader, base, cycles});
+    this._cacheResolvedContext({key: url, resolved, tag: remoteDoc.tag});
```

Comment:
Need some docs on the tag feature.  Is 'static' special?

Reply:
`static` is a special value, yes. We can document in a separate PR.